src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs

```diff
@@ -426,32 +427,20 @@ internal void CheckConstraintLanguageVersionAndRuntimeSupportForConversion(Synta
         {
             var syntax = (CSharpSyntaxNode)node.Syntax;
-            var implicitReceiver = new BoundObjectOrCollectionValuePlaceholder(syntax, isNewInstance: true, targetType) { WasCompilerGenerated = true };
             TypeSymbol? elementType;
-            BoundCollectionLiteralExpression collectionLiteral;
-
-            if (targetType is ArrayTypeSymbol arrayType)
-            {
-                if (!arrayType.IsSZArray)
-                {
-                    Error(diagnostics, ErrorCode.ERR_CollectionLiteralTargetTypeNotConstructible, syntax, targetType);
-                }
-                collectionLiteral = bindArrayOrSpan(syntax, spanConstructor: null, implicitReceiver, node.Initializers, arrayType.ElementType, diagnostics);
-            }
-            else if (isSpanType(targetType, WellKnownType.System_Span_T, out elementType))
-            {
-                var spanConstructor = getSpanConstructor(syntax, targetType, WellKnownMember.System_Span_T__ctor_Array, diagnostics);
-                collectionLiteral = bindArrayOrSpan(syntax, spanConstructor, implicitReceiver, node.Initializers, elementType, diagnostics);
-            }
-            else if (isSpanType(targetType, WellKnownType.System_ReadOnlySpan_T, out elementType))
-            {
-                var spanConstructor = getSpanConstructor(syntax, targetType, WellKnownMember.System_ReadOnlySpan_T__ctor_Array, diagnostics);
-                collectionLiteral = bindArrayOrSpan(syntax, spanConstructor, implicitReceiver, node.Initializers, elementType, diagnostics);
-            }
-            else
-            {
-                collectionLiteral = bindCollectionInitializer(syntax, implicitReceiver, node.Initializers, diagnostics);
-            }
+            BoundCollectionLiteralExpression collectionLiteral = Conversions.GetConstructibleCollectionType(Compilation, targetType, out elementType) switch
+            {
+                ConversionsBase.ConstructibleCollectionTypeKind.Array
+                    => bindArrayOrSpan(syntax, targetType, spanConstructor: null, node.Initializers, elementType!, diagnostics),
+                ConversionsBase.ConstructibleCollectionTypeKind.Span
+                    => bindArrayOrSpan(syntax, targetType, getSpanConstructor(syntax, targetType, WellKnownMember.System_Span_T__ctor_Array, diagnostics), node.Initializers, elementType!, diagnostics),
+                ConversionsBase.ConstructibleCollectionTypeKind.ReadOnlySpan
+                    => bindArrayOrSpan(syntax, targetType, getSpanConstructor(syntax, targetType, WellKnownMember.System_ReadOnlySpan_T__ctor_Array, diagnostics), node.Initializers, elementType!, diagnostics),
+                ConversionsBase.ConstructibleCollectionTypeKind.CollectionInitializer
+                    => BindCollectionInitializerCollectionLiteral(node, targetType, wasTargetTyped: true, wasCompilerGenerated: wasCompilerGenerated, diagnostics),
+                var collectionTypeKind
```

Comment:
No handling of `None`?

Reply:
`ConvertCollectionLiteralExpression()` is only called when the target type is a constructible collection type, so `None` is unexpected.